### ISSUE TITLE
Adding constraints blocks deployment

### ISSUE THREAD

**Issue by bobrik** [mesosphere/marathon#1133]
1. Create an app without any constraints, deploy it: everything looks fine.
2. Add `["hostname", "CLUSTER", "any-existing-node"]` constraint, deploy: deployment is stuck.
3. Remove app, deploy with the same constraint: everything looks fine again.

Also happens when `GROUP_BY` extended with `CLUSTER`, 0.8.0-SNAPSHOT. I remember seeing this since 0.7.3.

**Comment by dangra** [comment]
Still the same on 0.8.1

In my case the new constraint is like `["rack", "CLUSTER", "eu-1"]`. 
Marathon can't launch new tasks when its apps have existent tasks running in nodes that doesn't match the constraint. It doesn't make sense to me, it should relaunch new tasks in nodes matching the constraint and do not apply constraints to existent tasks.

**Comment by aquamatthias** [comment]
Hey @bobrik. The default `minimumHealthCapacity` if not defined in the application definition is 1. That means: start all instances of the new app definition, before killing instances of the old definition. The constraint says: a new instance can only be launched, if the old one goes away. Do we speak about that problem?

How to solve this:
- Specify a minimumHealthCapacity < 1 depending on the constraint. If this is not a problem, you can set it to 0. This is the behavior you describe by delete and add the application.
- If the minimumHealthCapacity is needed, think of the constraint and ensure it can be fulfilled 

I will leave that ticket open. Please close, if this solves the problem.

**Comment by bobrik** [comment]
@aquamatthias no, we speak about different things. I don't see why `minimumHealthCapacity` is relevant here since 2 tasks can be launched for an app without violating any constraints.

We hit the same bug on 0.10.0-RC3.

**Comment by aquamatthias** [comment]
@bobrik You are right - the problem I described is the unique constraint, but should not apply for CLUSTER or GROUP_BY constraints. I guess there are enough resources on the host `any-existing-node`, so that the constraint `["hostname", "CLUSTER", "any-existing-node"]` can be fulfilled?

**Comment by bobrik** [comment]
Looks like this time there were a mistake on our side (developer didn't push an image), my tests work on 0.10.0. Good to see it was fixed since 0.8.0. Sad to see that failures like that are still hard to debug.

Thanks!

**Comment by AlexClineBB** [comment]
I believe we also encountered this bug in v0.8.2. We added new constraints to an app definition that was already running and the deployment would stall during the RestartApplication phase.

    upgradeStrategy: {
        minimumHealthCapacity: 0.5
    }

By adding the above to the deployment definition, we were able to workaround the bug and keep half the number of instances running during the application restart. This only worked on apps running two or more tasks. Apps running a single task required the minimumHealthCapacity to be set to 0.
